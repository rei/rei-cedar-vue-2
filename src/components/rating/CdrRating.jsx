import clsx from 'clsx';
import contentPriority from '../../mixins/contentPriority';
import size from '../../mixins/size';
import space from '../../mixins/space';
import style from './styles/CdrRating.scss';

export default {
  name: 'CdrRating',
  mixins: [
    contentPriority,
    size,
    space,
  ],
  props: {
    /**
     * Rating value (out of 5)
     */
    rating: {
      required: true,
      type: [String, Number],
      default: 0,
    },
    /**
     * Total number of ratings
     */
    count: {
      required: false,
      type: [String, Number],
      default: null,
    },
    /**
     * Hides the word 'reviews' if true
     */
    compact: {
      type: Boolean,
      default: false,
    },
    /**
     * Allows the ratings to act as a link
     */
    href: {
      type: String,
    },
  },
  data() {
    return {
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-rating';
    },
    rounded() {
      return Math.round(this.rating * 4) / 4;
    },
    whole() {
      return Math.floor(this.rounded);
    },
    remainder() {
      return this.rounded.toFixed(2).split('.')[1];
    },
    empties() {
      return 5 - this.whole - (this.remainder > 0 ? 1 : 0);
    },
    formattedCount() {
      return this.compact ? `(${this.count})` : `${this.count}`;
    },
    remainderEl() {
      let remainder;
      if (this.remainder === '25') {
        remainder = (<span
          class={clsx(this.style['cdr-rating__icon'], this.style['cdr-rating__25'])}
          aria-hidden="true"
        />);
      } else if (this.remainder === '50') {
        remainder = (<span
          class={clsx(this.style['cdr-rating__icon'], this.style['cdr-rating__50'])}
          aria-hidden="true"
        />);
      } else if (this.remainder === '75') {
        remainder = (<span
          class={clsx(this.style['cdr-rating__icon'], this.style['cdr-rating__75'])}
          aria-hidden="true"
        />);
      }
      return remainder;
    },
    ratingSrText() {
      return `rated ${this.rounded} out of 5`;
    },
    countSrText() {
      return this.count !== null ? ` with ${this.count} reviews` : '';
    },
  },
  render() {
    const Component = this.href ? 'a' : 'div';
    return (
      <Component
        href={this.href}
        class={clsx(
          this.style[this.baseClass],
          this.contentPriorityClass,
          this.sizeClass,
          this.space,
          this.href ? this.style['cdr-rating--linked'] : '',
        )}
      >
        <div class={this.style['cdr-rating__ratings']}>

          {[...Array(this.whole).keys()].map(n => (
              <span
                class={clsx(this.style['cdr-rating__icon'], this.style['cdr-rating__100'])}
                key={`rating-whole-${n}-${this._uid}`} // eslint-disable-line no-underscore-dangle
                aria-hidden="true"
              />
          ))}
          {this.remainderEl }
          {[...Array(this.empties).keys()].map(n => (
            <span
              class={clsx(
                this.style['cdr-rating__icon'],
                (this.rounded > 0 || this.count > 0)
                  ? this.style['cdr-rating__placeholder']
                  : this.style['cdr-rating__placeholder--no-reviews'],
              )}
              key={`rating-empty-${n}-${this._uid}`} // eslint-disable-line no-underscore-dangle
              aria-hidden="true"
            />
          ))}
        </div>
        {this.count !== null
          ? <span
            aria-hidden="true"
            class={this.style['cdr-rating__count']}
          >
            {this.href
              && <span class={this.style['cdr-rating__number']}>
                { this.rounded }
              </span>
            }

            <span>
              { this.formattedCount }
            </span>

            {!this.compact
              && <span>
                &nbsp;Reviews
              </span>
            }
          </span> : ''
        }

        <span class={this.style['cdr-rating__caption-sr']}>
          { this.ratingSrText }{ this.countSrText }
        </span>
      </Component>
    );
  },
};
