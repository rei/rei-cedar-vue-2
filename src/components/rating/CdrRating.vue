<template>
  <!-- eslint-disable max-len -->
  <component
    :is="href ? 'a' : 'div'"
    :href="href"
    :class="[
      contentPriorityClass,
      sizeClass,
      space,
      href ? $style['cdr-rating--linked'] : '',
    ]"
  >
    <div :class="$style['cdr-rating__background']">
      <span
        :class="[
          $style['cdr-rating__icon'],
          count > 0 ? $style['cdr-rating__placeholder'] : $style['cdr-rating__placeholder--no-reviews'],
        ]"
        v-for="n in 5"
        :key="n"
        aria-hidden="true"
      />
    </div>
    <div :class="$style['cdr-rating__ratings']">
      <span
        :class="[$style['cdr-rating__icon'], $style['cdr-rating__100']]"
        v-for="n in whole"
        :key="n"
        aria-hidden="true"
      />
      <span
        v-if="remainder === '25'"
        :class="[$style['cdr-rating__icon'], $style['cdr-rating__25']]"
        aria-hidden="true"
      />
      <span
        v-else-if="remainder === '50'"
        :class="[$style['cdr-rating__icon'], $style['cdr-rating__50']]"
        aria-hidden="true"
      />
      <span
        v-else-if="remainder === '75'"
        :class="[$style['cdr-rating__icon'], $style['cdr-rating__75']]"
        aria-hidden="true"
      />
    </div>
    <span
      v-if="count"
      aria-hidden="true"
      :class="$style['cdr-rating__count']"
    ><span
      v-if="href"
      :class="$style['cdr-rating__number']"
    >{{ rounded }}</span><span>{{ formattedCount }}</span><span v-if="!compact">
      Reviews
    </span></span>
    <span
      class="cdr-sr-only"
    >rated {{ rounded }} out of 5 with {{ count }} reviews</span>
  </component>
</template>

<script>
import contentPriority from 'mixinsdir/contentPriority';
import size from 'mixinsdir/size';
import space from 'mixinsdir/space';
/**
 * Cedar 2 component for rating
 * Ratings are rounded to nearest .25 for displaying stars.
 * <span class="modifiers">Modifiers</span>
 * Modifiers can be combined 1 from each grouping.
 * {small, medium, large}
 * Default is large
 * @version 0.0.1
 * @author [REI Software Engineering](https://rei.github.io/rei-cedar/)
 */
export default {
  name: 'CdrRating',
  mixins: [
    contentPriority,
    size,
    space,
  ],
  props: {
    /**
     * Rating value (out of 5)
     */
    rating: {
      required: true,
      type: [String, Number],
      default: 0,
    },
    /**
     * Total number of ratings
     */
    count: {
      required: false,
      type: [String, Number],
      default: 0,
    },
    /**
     * Hides the word 'reviews' if true
     */
    compact: {
      type: Boolean,
      default: false,
    },
    /**
     * Allows the ratings to act as a link
     */
    href: {
      type: String,
    },
  },
  computed: {
    baseClass() {
      return 'cdr-rating';
    },
    rounded() {
      return Math.round(this.rating * 4) / 4;
    },
    whole() {
      return Math.floor(this.rounded);
    },
    remainder() {
      return this.rounded.toFixed(2).split('.')[1];
    },
    formattedCount() {
      return this.compact ? `(${this.count})` : `${this.count}`;
    },
  },
};
</script>

<style module>
  @import 'cssdir/settings/_index.pcss';
  @import './styles/vars/CdrRating.vars.pcss';
  @import './styles/CdrRating.pcss';
</style>
