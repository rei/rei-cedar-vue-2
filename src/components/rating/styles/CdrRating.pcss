/* ==========================================================================
  # CdrRating
  
  All values should map to variables in
  vars/CdrRating.vars.pcss
  in order to allow for theming

  TOC:

    :Base - Rating
      :Elements
        :Icon
        :Count
      :Size variants
========================================================================== */

.cdr-rating {
  display: flex;
  align-items: center;
  position: relative;

  /* Elements
    ========================================================================== */

  /* Icon
    ========== */
  &__icon {
    width: $ratings-icon-width;
    height: $ratings-icon-height;
  }

  /* Count
    ========== */
  &__count {
    margin-left: $ratings-count-margin-left;
    font-size: $ratings-count-font-size;
    line-height: $ratings-count-line-height;
    text-align: $ratings-count-text-align;
    font-style: $ratings-count-font-style;
    font-weight: $ratings-count-font-weight;
    font-family: $ratings-count-font-family;
    color: $ratings-count-color;
  }

  &__ratings {
    display: flex;
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    left: 0;
  }

  &__background {
    display: flex;
  }

  /* Placeholder
    ========== */

  &__placeholder {
    background-image: svg-load('../../../assets/icons/rei/star-100.svg', fill=$(ratings-placeholder-fill));
    background-repeat: no-repeat;
  }

  /* Stars
    ========== */

  &__100 {
    background-image: svg-load('../../../assets/icons/rei/star-100.svg', fill=$(ratings-star-fill));
    background-repeat: no-repeat;
  }

  &__75 {
    background-image: svg-load('../../../assets/icons/rei/star-75.svg', fill=$(ratings-star-fill));
    background-repeat: no-repeat;
  }

  &__50 {
    background-image: svg-load('../../../assets/icons/rei/star-50.svg', fill=$(ratings-star-fill));
    background-repeat: no-repeat;
  }

  &__25 {
    background-image: svg-load('../../../assets/icons/rei/star-25.svg', fill=$(ratings-star-fill));
    background-repeat: no-repeat;
  }

  /* Size variants
    ========================================================================== */

  /* Large
    ========== */

  &--large {
    & .cdr-rating__icon {
      width: $ratings-icon-large-width;
      height: $ratings-icon-large-height;
    }

    & .cdr-rating__count {
      margin-left: $ratings-count-large-margin-left;
      font-size: $ratings-count-large-font-size;
      line-height: $ratings-count-large-line-height;
      text-align: $ratings-count-large-text-align;
      font-style: $ratings-count-large-font-style;
      font-weight: $ratings-count-large-font-weight;
      font-family: $ratings-count-large-font-family;
      color: $ratings-count-large-color;
    }
  }

  /* Small
    ========== */

  &--small {
    & .cdr-rating__icon {
      width: $ratings-icon-small-width;
      height: $ratings-icon-small-height;
    }

    & .cdr-rating__count {
      margin-left: $ratings-count-small-margin-left;
      font-size: $ratings-count-small-font-size;
      line-height: $ratings-count-small-line-height;
      text-align: $ratings-count-small-text-align;
      font-style: $ratings-count-small-font-style;
      font-weight: $ratings-count-small-font-weight;
      font-family: $ratings-count-small-font-family;
      color: $ratings-count-small-color;
    }
  }
}
