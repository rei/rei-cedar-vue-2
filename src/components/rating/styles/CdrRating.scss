@import '../../../css/settings/index.scss';

$rating-color: #2b6692;

@mixin cdr-rating-small() {
  & .cdr-rating__icon {
    width: $cdr-icon-size-sm;
    height: $cdr-icon-size-sm;
  }

  & .cdr-rating__count {
    @include cdr-text-utility-sans-100;

    margin-left: $cdr-space-quarter-x;
  }
}

@mixin cdr-rating-medium($parent-selector: '& ') {
  #{$parent-selector}.cdr-rating__icon {
    width: $cdr-icon-size;
    height: $cdr-icon-size;
  }

  #{$parent-selector}.cdr-rating__count {
    @include cdr-text-utility-sans-200;

    margin-left: $cdr-space-half-x;
  }
}

@mixin cdr-rating-large() {
  & .cdr-rating__icon {
    width: $cdr-icon-size-lg;
    height: $cdr-icon-size-lg;
  }

  & .cdr-rating__count {
    @include cdr-text-utility-sans-300;
  }
}

@include cdr-rating-medium('');

.cdr-rating {
  display: flex;
  align-items: center;
  position: relative;
  line-height: 1;

  color: $cdr-color-text-rating-default;

  /* Elements
    ========================================================================== */

  &__number {
    font-weight: 600;
  }

  &__ratings {
    display: flex;
  }

  &__caption-sr {
    @include cdr-display-sr-only-mixin;
  }

  /* Placeholder
    ========== */

  &__placeholder {
    background-image: svg-load('node_modules/@rei/cedar-icons/dist/icons/star-stroke.svg', fill=#{$rating-color});
    background-repeat: no-repeat;

    &--no-reviews {
      background-image: svg-load('node_modules/@rei/cedar-icons/dist/icons/star-stroke.svg', fill=#{$cdr-color-text-form-placeholder-lightmode});
    }
  }

  /* Stars
    ========== */

  &__100 {
    background-image: svg-load('node_modules/@rei/cedar-icons/dist/icons/star-100.svg', fill=#{$rating-color});
    background-repeat: no-repeat;
  }

  &__75 {
    background-image: svg-load('node_modules/@rei/cedar-icons/dist/icons/star-75.svg', fill=#{$rating-color});
    background-repeat: no-repeat;
  }

  &__50 {
    background-image: svg-load('node_modules/@rei/cedar-icons/dist/icons/star-50.svg', fill=#{$rating-color});
    background-repeat: no-repeat;
  }

  &__25 {
    background-image: svg-load('node_modules/@rei/cedar-icons/dist/icons/star-25.svg', fill=#{$rating-color});
    background-repeat: no-repeat;
  }

  /* Modifiers
    ========================================================================== */

  /* Linked
    ========== */

  &--linked {
    display: inline-flex;
    text-decoration: none;

    & .cdr-rating__count {
      &:hover,
      &:active,
      &:focus {
        color: $cdr-color-text-rating-hover;
        text-decoration: underline;
      }

      & > .cdr-rating__number {
        border-right: 1px solid $cdr-color-text-rating-separator;
        padding: 0 $cdr-space-quarter-x;
        margin-right: $cdr-space-quarter-x;
      }
    }
  }

  /* Large
    ========== */

  &--large {
    @include cdr-rating-large;
  }

  /* Medium
    ========== */

  &--medium {
    @include cdr-rating-medium();
  }

  /* Small
    ========== */

  &--small {
    @include cdr-rating-small;
  }

  @include xs-mq-only {
    &--small\@xs {
      @include cdr-rating-small;
    }

    &--medium\@xs {
      @include cdr-rating-medium;
    }

    &--large\@xs {
      @include cdr-rating-large;
    }
  }

  @include sm-mq-only {
    &--small\@sm {
      @include cdr-rating-small;
    }

    &--medium\@sm {
      @include cdr-rating-medium;
    }

    &--large\@sm {
      @include cdr-rating-large;
    }
  }

  @include md-mq-only {
    &--small\@md {
      @include cdr-rating-small;
    }

    &--medium\@md {
      @include cdr-rating-medium;
    }

    &--large\@md {
      @include cdr-rating-large;
    }
  }

  @include lg-mq-only {
    &--small\@lg {
      @include cdr-rating-small;
    }

    &--medium\@lg {
      @include cdr-rating-medium;
    }

    &--large\@lg {
      @include cdr-rating-large;
    }
  }
}
