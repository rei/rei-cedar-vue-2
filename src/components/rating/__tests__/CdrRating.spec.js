import { shallowMount, mount } from '@vue/test-utils';
import CdrRating from 'componentdir/rating/CdrRating';

describe('CdrRating', () => {
  test('renders correctly', () => {
    const wrapper = mount(CdrRating, {
      propsData: {
        rating: 3.2323,
        count: 100,
      }
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('computes and rounds rating stars correctly', () => {
    const wrapper = shallowMount(CdrRating, {
      propsData: {
        rating: 3.4441231,
        count: 1,
      }
    });
    expect(wrapper.vm.whole).toBe(3);
    expect(wrapper.vm.remainder).toBe('50');
    expect(wrapper.vm.rounded).toBe(3.5);

    wrapper.setProps({rating: 3.122227});
    expect(wrapper.vm.whole).toBe(3);
    expect(wrapper.vm.remainder).toBe('00');
    expect(wrapper.vm.rounded).toBe(3);

    wrapper.setProps({rating: 3.222227});
    expect(wrapper.vm.whole).toBe(3);
    expect(wrapper.vm.remainder).toBe('25');
    expect(wrapper.vm.rounded).toBe(3.25);

    wrapper.setProps({rating: 3.673323});
    expect(wrapper.vm.whole).toBe(3);
    expect(wrapper.vm.remainder).toBe('75');
    expect(wrapper.vm.rounded).toBe(3.75);
  });

  it('computes and rounds rating stars correctly with string input', () => {
    const wrapper = shallowMount(CdrRating, {
      propsData: {
        rating: '3.4441231',
        count: '1',
      }
    });
    expect(wrapper.vm.whole).toBe(3);
    expect(wrapper.vm.remainder).toBe('50');
    expect(wrapper.vm.rounded).toBe(3.5);
  });

  it('renders an anchor when given an href', () => {
    const wrapper = shallowMount(CdrRating, {
      propsData: {
        rating: 5,
        href: 'rei.com'
      }
    });
    expect(wrapper.is('a')).toBeTruthy();
  });

  it('renders normal placeholder stars when count is > 0', () => {
    const wrapper = shallowMount(CdrRating, {
      propsData: {
        rating: 5,
        count: 1,
      }
    });
    expect(wrapper.contains('.cdr-rating__placeholder')).toBe(true);
    expect(wrapper.contains('.cdr-rating__placeholder--no-reviews')).toBe(false);
  });

  it('renders custom placeholder stars when count is 0', () => {
    const wrapper = shallowMount(CdrRating, {
      propsData: {
        rating: 5,
        count: 0,
        href: 'rei.com'
      }
    });
    expect(wrapper.contains('.cdr-rating__placeholder')).toBe(false);
    expect(wrapper.contains('.cdr-rating__placeholder--no-reviews')).toBe(true);
  });

  it('renders custom placeholder stars when count is "0"', () => {
    const wrapper = shallowMount(CdrRating, {
      propsData: {
        rating: 5,
        count: "0",
        href: 'rei.com'
      }
    });
    expect(wrapper.contains('.cdr-rating__placeholder')).toBe(false);
    expect(wrapper.contains('.cdr-rating__placeholder--no-reviews')).toBe(true);
  });
});
