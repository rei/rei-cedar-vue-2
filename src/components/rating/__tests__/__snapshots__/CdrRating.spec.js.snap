// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrRating renders correctly 1`] = `
<div
  class="cdr-rating cdr-rating--primary cdr-rating "
>
  <div
    class="cdr-rating__background"
  >
    <span
      aria-hidden="true"
      class="cdr-rating__icon cdr-rating__placeholder"
    />
    <span
      aria-hidden="true"
      class="cdr-rating__icon cdr-rating__placeholder"
    />
    <span
      aria-hidden="true"
      class="cdr-rating__icon cdr-rating__placeholder"
    />
    <span
      aria-hidden="true"
      class="cdr-rating__icon cdr-rating__placeholder"
    />
    <span
      aria-hidden="true"
      class="cdr-rating__icon cdr-rating__placeholder"
    />
  </div>
  <div
    class="cdr-rating__ratings"
  >
    <span
      aria-hidden="true"
      class="cdr-rating__icon cdr-rating__100"
    />
    <span
      aria-hidden="true"
      class="cdr-rating__icon cdr-rating__100"
    />
    <span
      aria-hidden="true"
      class="cdr-rating__icon cdr-rating__100"
    />
    <span
      aria-hidden="true"
      class="cdr-rating__icon cdr-rating__25"
    />
  </div>
  <span
    aria-hidden="true"
    class="cdr-rating__count"
  >
    <span>
      100
    </span>
    <span>
      Â Reviews
    </span>
  </span>
  <span
    class="cdr-display-sr-only"
  >
    rated 3.25 out of 5 with 100 reviews
  </span>
</div>
`;
